Add support for customizing the catalog graph relations using an API

grantila opened the pull request:
## Better support custom relations in the catalog graph plugin

The catalog graph plugin currently only understands the built-in relations with [this caveat](https://github.com/backstage/backstage/blob/aedd4daa661c57b6067d872f67884af461a710b8/plugins/catalog-graph/src/components/EntityRelationsGraph/relations.ts#L41-L44). It will however by default show any other relation an entity has.

This has drawbacks:
 * The UI won't autocomplete for other "unknown" relations
 * The UI will show relations that might not be suitable to show
 * Configuring this is needed everywhere the graph is shown (its own page, the entity pages, etc)

This MR adds an API to the catalog graph plugin, which can be customized. It allows an implementor to specify exactly what relations and relation pairs to support, and what of them to show/select by default. The plugin will use this configuration wherever the graph is shown (although overridable by its already existing props).

The default implementation of the API includes the same known relations as today, so should be no different than before.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

grantila commented on .changeset/dull-eagles-sin.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-catalog-graph': patch
```
Thanks, yeah that's fair, changed.

vinzscam reviewed (commented):
Thanks @grantila, it looks great! 👍

Do you think this could be the right place for addressing https://github.com/backstage/backstage/issues/30121?

grantila commented:
Yeah this PR would be a great start. I'm actively working on solving #30121 (which currently is based on the logic in this branch) and will have a fix ready shortly (this week most likely).

freben commented:
I rudely pushed a fix.

Not sure why this would differ, maybe we're on very very slightly different versions of something that affected the api reports. They are a little finicky. But this is what I got after just grabbing your branch and running `yarn` and `yarn build:api-reports`.

grantila commented:
Thanks @freben, the opposite of rude 😄

If we get this in, I'll try as quickly as possible to fix the "simplify" (internally called "unidirectional", although not being the exact same thing 😏) which isn't working anymore. 🙏

grantila commented:
@freben & @vinzscam I now have a fix for #30121. Before realizing the issue of 30121, I started refactoring the logic heavily, and implemented a `catalog-graph-backend` to massively speed up the load/rendering, but I have extracted the pure UI logic/fixes (to not introduce a new required backend just for a bugfix). Will provide a followup PR for the backend later, so it can introduce a breaking change (requiring the backend plugin).

However, it's a pretty big refactoring of the graph logic (lots of changes), and I have compared it side-by-side heavily with a branch before #29895 when the bug was introduced, with various graphs of various sizes with combinations of `mergeRelations`/`unidirectional`. It's based on this branch, so I could either commit it here, or wait for this and have a separate PR (if that's easier to grasp).
What's your input on this, to have it solved as quickly as possible?

grantila commented:
After fixing missing edges in #29895, the "simplified" didn't work anymore (#30121). The graph logic was quite hard to grasp, and before knowing about 30121, I started refactoring it, upon which this fix was made.

The graph logic is extracted into `/lib/graph` (for constructing the graph from a set of entities) and `/lib/graph-transformations` where each transformation is standalone and more easy understand.

The [useEntityRelationNodesAndEdges.ts](https://github.com/backstage/backstage/pull/30889/commits/f60e9fc1f8ac30474ecf7cbaa305560b4d1dea91#diff-65b94eef8151c7948cc461c41801aefcc6b07134db6066257ac82a3fb347bf71) is now smaller and quite clear what it does.

With moving towards a generic interface for transforming the graph, we could eventually introduce and allow for custom transformers to be supplied as a prop to the graph, so implementors can transform the graph to whatever need they have. Defaulting to the built-in ones will be necessary for exact backwards compatibility though.

vinzscam commented on plugins/catalog-graph/src/components/EntityRelationsGraph/useEntityRelationGraph.ts:
```diff
@@ -98,19 +100,21 @@ export function useEntityRelationGraph({
-        : entities,
-    [entities, entityFilter],
-  );
+  const filteredEntities = useMemo(() => {
+    if (loading) {
+      return {};
+    }
+
+    return entityFilter
+      ? pickBy(entities, (value, _key) => entityFilter(value))
+      : entities;
+  }, [loading, entities, entityFilter]);
```
not sure if it could make sense to ignore the loading state, so that we avoid some flickering if the data is reloaded. Feel free to ignore if this introduces side effects 

```suggestion
  const filteredEntities = useMemo(() => {
    return entityFilter
      ? pickBy(entities, (value, _key) => entityFilter(value))
      : entities;
  }, [entities, entityFilter]);
```

grantila replied:
Ironically, it'll flicker less now.
The upstream `entities` will contain a bunch of entities. When the user changes the query, a new set of entities will be fetched (and `loading` will be true), but the old ones will remain in the `entities`, so leaving the UI with a progress spinner on top flickers less than "emptying" it with this `if (loading)` ☺️

vinzscam commented on .changeset/dull-eagles-sin.md:
```diff
@@ -0,0 +1,6 @@
+---
+
+Support custom relations by using an API to define known relations and which to show by default
+Fixes "simplified" bug (#30121) which caused graphs not to be simplified
```
✨

vinzscam reviewed (approved):
🚀

freben reviewed (commented):
Alright, this is nice.

There's a ton of spread->`Set`->`Array.from` and `array.some` and stuff like that. I wonder if there's ever going to be reports of "i tried opening the graph page for my entity with 2000 relations and the tab crashed".

freben commented on plugins/catalog-graph/src/components/EntityRelationsGraph/EntityRelationsGraph.tsx:
```diff
@@ -91,6 +92,7 @@ export type EntityRelationsGraphProps = {
   renderLabel?: DependencyGraphTypes.RenderLabelFunction<EntityEdge>;
   curve?: 'curveStepBefore' | 'curveMonotoneX';
   showArrowHeads?: boolean;
+  onPostTransformation?: GraphTransformationDebugger;
```
Yepyep.

In general, debugging stuff shouldn't be part of public interfaces.

grantila commented:
After feedback; removed the `onPostTransformation` prop and including debug related logic. cc @freben
